node/src/env0-deploy-utils.js

```diff
@@ -54,40 +52,94 @@ class DeployUtils {
   }
 
   async deployEnvironment(environment, blueprintRevision, blueprintId) {
-    console.log(`starting to deploy environmentId: ${environment.id}, blueprintId: ${blueprintId}`);
-    const deployment = await apiClient.callApi('post', `environments/${environment.id}/deployments`,
-        {data: {blueprintId, blueprintRevision}});
+    await this.waitForEnvironment(environment.id);
 
-    console.log(`Started deployment ${deployment.id}`);
+    return await apiClient.callApi('post', `environments/${environment.id}/deployments`, {data: {blueprintId, blueprintRevision}});
   }
 
   async destroyEnvironment(environment) {
-    console.log(`Starting to destroy environmentId: ${environment.id}`);
-    const deployment = await apiClient.callApi('post', `environments/${environment.id}/destroy`);
-    console.log(`Started destroy ${deployment.id}`);
+    await this.waitForEnvironment(environment.id);
+
+    return await apiClient.callApi('post', `environments/${environment.id}/destroy`);
+  }
+
+  async writeDeploymentStepLog(deploymentLogId, stepName) {
+    let hasMoreLogs;
+
+    do {
+      const { events, hasMoreLogs: currentHasMoreLogs } = await apiClient.callApi('get', `deployments/${deploymentLogId}/steps/${stepName}/log`);
+      events.forEach((event) => console.log(event.message));
+
+      hasMoreLogs = currentHasMoreLogs;
+      await apiClient.sleep(1000);
+    } while (hasMoreLogs)
   }
 
+  async fetchDeploymentSteps(deploymentLogId, stepsToSkip) {
+    const inProgressStepStatuses = ['IN_PROGRESS', 'NOT_STARTED'];
+    const doneSteps = [];
+
+    const steps = await apiClient.callApi('get', `deployments/${deploymentLogId}/steps`);
+
+    for (const step of steps) {
+      const alreadyLogged = stepsToSkip.includes(step.name);
+      const inProgress = inProgressStepStatuses.includes(step.status);
+
+      if (!alreadyLogged && !inProgress) {
+        console.log(`$$$ ${step.name}`);
+
+        console.log(range(100).map(() => '#').join(''));
+        await this.writeDeploymentStepLog(deploymentLogId, step.name);
+
+        doneSteps.push(step.name);
+      }
+    }
 
-  async pollEnvironmentStatus(environmentId) {
-    const deployEndStatuses = ['CREATED', 'ACTIVE', 'INACTIVE', 'FAILED', 'TIMEOUT'];
-    const maxRetryNumber = 90; //waiting for 15 minutes (90 * 10 seconds)
+    return doneSteps;
+  }
+
+  async pollDeploymentStatus(deploymentLogId) {
+    let stepsAlreadyLogged = [];
+    const maxRetryNumber = 200;
```

Comment:
Changed the `pollDeploymentStatus` to be by time diff and not by retries calculation (since nested functions also use `sleep`). Set it to 3 hours (If a user waits for more than 3 hours on CLI then he deserves to get a timeout 😸 )
6626e5e

Reply:
Awesome, way better